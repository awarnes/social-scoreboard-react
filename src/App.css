a, a:active, a:hover, a:visited, a:link {
  text-decoration: none;
  color: black;
}
