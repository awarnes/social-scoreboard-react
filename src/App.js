import React, { Component } from 'react'
import {
  BrowserRouter as Router,
  Route,
  Redirect
} from 'react-router-dom'
import { Well } from 'react-bootstrap'
import fire, { auth, provider } from './firebaseConfig'

import './App.css'
import CreateBoard from './CreateBoard'
import ScoreBoard from './ScoreBoard'
import LoginPage from './LoginPage'
import UserDashboard from './UserDashboard'

class App extends Component {
  constructor (props) {
    super(props)

    this.state = {
      contestantNames: '',
      boardInfo: {
        boardTitle: '',
        boardContestants: []
      },
      activeBoardId: '',
      user: null,
      userName: ''
    }

    this.updateBoardTitle = this.updateBoardTitle.bind(this)
    this.updateBoardContestants = this.updateBoardContestants.bind(this)
    this.incrementScore = this.incrementScore.bind(this)
    this.decrementScore = this.decrementScore.bind(this)
    this.incrementAll = this.incrementAll.bind(this)
    this.decrementAll = this.decrementAll.bind(this)
    this.clearAll = this.clearAll.bind(this)
    this.createNewBoard = this.createNewBoard.bind(this)
    this.updateScoreBoardFromDatabase = this.updateScoreBoardFromDatabase.bind(this)
    this.login = this.login.bind(this)
    this.logout = this.logout.bind(this)
    this.getUserInformation = this.getUserInformation.bind(this)
  }

  componentDidMount () {
    auth.onAuthStateChanged((user) => {
      if (user) {
        this.setState({ user, userName: user.displayName }, () => { this.getUserInformation() })
      }
    })
  }

  updateBoardTitle (evt) {
    const newBoardInfo = Object.assign({}, this.state.boardInfo)
    newBoardInfo.boardTitle = evt.target.value
    this.setState({boardInfo: newBoardInfo})
  }

  updateBoardContestants (evt) {
    const contestantNames = evt.target.value.split(',')

    const newBoardInfo = Object.assign({}, this.state.boardInfo)

    const boardContestants = contestantNames.map((name) => {
      return {name: name.trim(), score: 0}
    })

    newBoardInfo.boardContestants = boardContestants

    this.setState({boardInfo: newBoardInfo, contestantNames: evt.target.value})
  }

  incrementScore (incContestant) {
    const newBoardInfo = Object.assign({}, this.state.boardInfo)

    newBoardInfo.boardContestants.forEach((contestant) => {
      if (contestant.name === incContestant) {
        contestant.score += 1
      }
    })

    this.setState({boardInfo: newBoardInfo}, () => { this.updateScoresInDatabase() })
  }

  decrementScore (decContestant) {
    const newBoardInfo = Object.assign({}, this.state.boardInfo)

    newBoardInfo.boardContestants.forEach((contestant) => {
      if (contestant.name === decContestant) {
        contestant.score -= 1
      }
    })

    this.setState({boardInfo: newBoardInfo}, () => { this.updateScoresInDatabase() })
  }

  incrementAll () {
    const newBoardInfo = Object.assign({}, this.state.boardInfo)

    newBoardInfo.boardContestants.forEach((contestant) => {
      contestant.score += 1
    })

    this.setState({boardInfo: newBoardInfo}, () => { this.updateScoresInDatabase() })
  }

  decrementAll () {
    const newBoardInfo = Object.assign({}, this.state.boardInfo)

    newBoardInfo.boardContestants.forEach((contestant) => {
      contestant.score -= 1
    })

    this.setState({boardInfo: newBoardInfo}, () => { this.updateScoresInDatabase() })
  }

  clearAll () {
    const newBoardInfo = Object.assign({}, this.state.boardInfo)

    newBoardInfo.boardContestants.forEach((contestant) => {
      contestant.score = 0
    })

    this.setState({boardInfo: newBoardInfo}, () => { this.updateScoresInDatabase() })
  }

  updateScoresInDatabase () {
    if (this.state.activeBoardId !== '') {
      const boardInfoRef = fire.database().ref(`boards/${this.state.activeBoardId}`)
      boardInfoRef.set(this.state.boardInfo)
    }
  }

  createNewBoard () {
    const boardsRef = fire.database().ref('boards')
    boardsRef.push(this.state.boardInfo)
    boardsRef.once('child_added')
      .then((dataSnapshot) => {
        this.setState({activeBoardId: dataSnapshot.key})
      })
  }

  updateScoreBoardFromDatabase (boardUid) {
    const boardInfoRef = fire.database().ref(`boards/${boardUid}`)
    boardInfoRef.once('value')
      .then((snapshot) => {
        let boardInfo = snapshot.val()
        this.setState({boardInfo, activeBoardId: boardUid})
      })
  }

  getUserInformation () {
    const userRef = fire.database().ref(`users/${this.state.user.uid}`)
    userRef.once('child_added')
      .then((snapshot) => {
        console.log(snapshot)
      })
  }

  createUser () {
    const userRef = fire.database().ref(`users/${this.state.user.uid}`)
    userRef.once('value')
      .then((dataSnapshot) => {
        console.log('dataSnapshot', dataSnapshot)
        if (!dataSnapshot.exists()) {
          console.log('dataSnapshot.key', dataSnapshot.key)
        }
      })
  }

  login () {
    auth.signInWithPopup(provider)
      .then((result) => {
        console.log('result', result)
        const user = result.user
        console.log('user', user)
        this.setState({user}, () => { this.createUser() })
      })
  }

  logout () {
    auth.signOut()
      .then(() => {
        this.setState({user: null})
      })
  }

  render () {
    return (
      <Well>
        <Router>
          <div>
            <Route exact path='/'
              render={(props) =>
                (this.state.user ? (<Redirect to={`/${this.state.user.uid}/dashboard`} />)
                  : (<LoginPage {...props}
                    login={this.login}
                  />))} />

            <Route path={`/:uid/dashboard`}
              render={props => (<UserDashboard {...props}
                logout={this.logout}
                userName={this.state.userName}
              />)} />

            <Route path='/createboard'
              render={props => (<CreateBoard {...props}
                updateBoardTitle={this.updateBoardTitle}
                updateBoardContestants={this.updateBoardContestants}
                boardTitle={this.state.boardInfo.boardTitle}
                contestantNames={this.state.contestantNames}
                boardContestants={this.state.boardInfo.boardContestants}
                createNewBoard={this.createNewBoard}
                activeBoardId={this.state.activeBoardId}
              />)} />
            <Route path='/board/:uid'
              render={props => (<ScoreBoard {...props}
                updateScoreBoardFromDatabase={this.updateScoreBoardFromDatabase}
                boardInfo={this.state.boardInfo}
                incrementScore={this.incrementScore}
                decrementScore={this.decrementScore}
                incrementAll={this.incrementAll}
                decrementAll={this.decrementAll}
                clearAll={this.clearAll}
              />)} />
          </div>
        </Router>
      </Well>
    )
  }
}

export default App
