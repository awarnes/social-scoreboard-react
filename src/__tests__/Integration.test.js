import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'

import App from '../App'

/* global it describe expect */

describe('App', () => {
  it('renders according to the snapshot', () => {
    const tree = renderer.create(<App />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(<App />, div)
  })
})
