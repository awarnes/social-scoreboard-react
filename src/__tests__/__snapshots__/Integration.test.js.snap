// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders according to the snapshot 1`] = `
<div
  className="well"
>
  <div>
    <div
      className="row"
    >
      <div
        className="col-xs-3"
      />
      <div
        className="col-xs-6"
      >
        <div
          className="row"
        >
          <div
            className="col-xs-1"
          />
          <div
            className="col-xs-10"
          >
            <h1>
              Create a ScoreBoard!
            </h1>
          </div>
          <div
            className="col-xs-1"
          />
        </div>
        <div
          className="row"
        >
          <div
            className="col-xs-12"
          >
            <div
              className="form-group"
            >
              <label
                className="control-label"
                htmlFor="titleInput"
                id="titleLabel"
              >
                Scoreboard Title:
              </label>
              <input
                className="form-control"
                id="titleInput"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="col-xs-12"
          >
            <div
              className="form-group"
            >
              <label
                className="control-label"
                htmlFor="contestantInput"
                id="contestantLabel"
              >
                Contestants:
              </label>
              <input
                className="form-control"
                id="contestantInput"
                onChange={[Function]}
                type="text"
                value=""
              />
              <span
                className="help-block"
              >
                Please enter a comma seperated list.
              </span>
            </div>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="col-xs-4"
          />
          <div
            className="col-xs-4"
          >
            <button
              className="btn btn-default"
              disabled={false}
              type="button"
            >
              <a
                href="/board"
                onClick={[Function]}
              >
                Create Board
              </a>
            </button>
          </div>
          <div
            className="col-xs-4"
          />
        </div>
      </div>
      <div
        className="col-xs-3"
      />
    </div>
  </div>
</div>
`;
